README.md
# Helix Redirect

> Serve 3xx and internal redirects for Project Helix

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/helix-redirect.svg)](https://codecov.io/gh/adobe/helix-redirect)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/helix-redirect.svg)](https://circleci.com/gh/adobe/helix-redirect)
[![GitHub license](https://img.shields.io/github/license/adobe/helix-redirect.svg)](https://github.com/adobe/helix-redirect/blob/main/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/helix-redirect.svg)](https://github.com/adobe/helix-redirect/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/helix-redirect.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/helix-redirect)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release) 

## Installation

## Usage

`helix-redirect` is a service that will be called from `helix-dispatch`. It reads a repository's [`helix-redirects.yaml`](https://github.com/adobe/helix-shared/blob/main/docs/redirects.md), retrieves all linked redirect spreadsheets, evaluates the current URL against the configured redirect rules and returns either a 3xx redirect or a 204 (no content) response to indicate that no redirect rules apply to the current URL.

```bash
curl https://adobeioruntime.net/api/v1/web/helix/helix-services/redirect@v1
```

For more, see the [API documentation](docs/API.md).

## Development

### Deploying Helix Redirect

Deploying Helix Redirect requires the `wsk` command line client, authenticated to a namespace of your choice. For Project Helix, we use the `helix` namespace.

All commits to main that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-services/redirect@ci<num>` and tagged with the CI build number.


CHANGELOG.md
## [1.4.71](https://github.com/adobe/helix-redirect/compare/v1.4.70...v1.4.71) (2022-02-23)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.7.15 ([a8c909f](https://github.com/adobe/helix-redirect/commit/a8c909fcb2cc4dc837ea28e25056ce2c24a1a403))

## [1.4.70](https://github.com/adobe/helix-redirect/compare/v1.4.69...v1.4.70) (2022-02-21)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.2.7 ([430fc60](https://github.com/adobe/helix-redirect/commit/430fc60dc82313cb49adae404fcab32e10c33715))

## [1.4.69](https://github.com/adobe/helix-redirect/compare/v1.4.68...v1.4.69) (2022-02-21)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.7.14 ([2174983](https://github.com/adobe/helix-redirect/commit/2174983c89fb334759a8bea73547b843dbfd27bf))

## [1.4.68](https://github.com/adobe/helix-redirect/compare/v1.4.67...v1.4.68) (2022-02-11)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.7.9 ([#419](https://github.com/adobe/helix-redirect/issues/419)) ([867e483](https://github.com/adobe/helix-redirect/commit/867e483e44954210ab7c7c5510aeeb1a5838f016))

## [1.4.67](https://github.com/adobe/helix-redirect/compare/v1.4.66...v1.4.67) (2022-02-09)


### Bug Fixes

* **deps:** update adobe fixes ([#418](https://github.com/adobe/helix-redirect/issues/418)) ([c585c9d](https://github.com/adobe/helix-redirect/commit/c585c9d84bf29ec350c55c6b4d36be874b6abf58))

## [1.4.66](https://github.com/adobe/helix-redirect/compare/v1.4.65...v1.4.66) (2022-02-06)


### Bug Fixes

* **deps:** update adobe fixes ([#410](https://github.com/adobe/helix-redirect/issues/410)) ([0838344](https://github.com/adobe/helix-redirect/commit/08383446dc9a4c725a6798e9132b3d69dda67d5f))

## [1.4.65](https://github.com/adobe/helix-redirect/compare/v1.4.64...v1.4.65) (2022-02-04)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.2.5 ([#408](https://github.com/adobe/helix-redirect/issues/408)) ([86e5804](https://github.com/adobe/helix-redirect/commit/86e5804df604f2d6115f366012f3cc4efca93249))

## [1.4.64](https://github.com/adobe/helix-redirect/compare/v1.4.63...v1.4.64) (2022-02-03)


### Bug Fixes

* **deps:** update adobe fixes ([#406](https://github.com/adobe/helix-redirect/issues/406)) ([b710a33](https://github.com/adobe/helix-redirect/commit/b710a33a355fd3e265604b8b4095a1021941bc66))

## [1.4.63](https://github.com/adobe/helix-redirect/compare/v1.4.62...v1.4.63) (2022-01-27)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.7.5 ([#398](https://github.com/adobe/helix-redirect/issues/398)) ([164dcf7](https://github.com/adobe/helix-redirect/commit/164dcf7d1a97b8d7728196419547b7864d8324da))

## [1.4.62](https://github.com/adobe/helix-redirect/compare/v1.4.61...v1.4.62) (2022-01-10)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v2 ([#386](https://github.com/adobe/helix-redirect/issues/386)) ([c138fd7](https://github.com/adobe/helix-redirect/commit/c138fd7953a23dac2e6ad40a4ee57a589db387a8))

## [1.4.61](https://github.com/adobe/helix-redirect/compare/v1.4.60...v1.4.61) (2021-12-07)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.7.1 ([#376](https://github.com/adobe/helix-redirect/issues/376)) ([80733bb](https://github.com/adobe/helix-redirect/commit/80733bb898c447e356bf34e78fbb521f4e889eb9))

## [1.4.60](https://github.com/adobe/helix-redirect/compare/v1.4.59...v1.4.60) (2021-12-04)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.7.0 ([#373](https://github.com/adobe/helix-redirect/issues/373)) ([856fdcb](https://github.com/adobe/helix-redirect/commit/856fdcb7d1c78a9676835f23ce5dbe184ad0893e))

## [1.4.59](https://github.com/adobe/helix-redirect/compare/v1.4.58...v1.4.59) (2021-12-03)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.6.0 ([#370](https://github.com/adobe/helix-redirect/issues/370)) ([289c704](https://github.com/adobe/helix-redirect/commit/289c704f15b2cd18720d06e9e6f3c162251b3901))

## [1.4.58](https://github.com/adobe/helix-redirect/compare/v1.4.57...v1.4.58) (2021-11-29)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.5.3 ([#369](https://github.com/adobe/helix-redirect/issues/369)) ([65f00fb](https://github.com/adobe/helix-redirect/commit/65f00fbb84adbc63cf22abbc86e4fdd7201ee8d6))

## [1.4.57](https://github.com/adobe/helix-redirect/compare/v1.4.56...v1.4.57) (2021-11-26)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.9.1 ([#367](https://github.com/adobe/helix-redirect/issues/367)) ([1a5c26b](https://github.com/adobe/helix-redirect/commit/1a5c26bb0dd955b180a7fa51219ddd6ba71002f1))

## [1.4.56](https://github.com/adobe/helix-redirect/compare/v1.4.55...v1.4.56) (2021-11-12)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.5.0 ([#360](https://github.com/adobe/helix-redirect/issues/360)) ([807afa8](https://github.com/adobe/helix-redirect/commit/807afa896dbd919cf717b44ba42a4282aef96392))

## [1.4.55](https://github.com/adobe/helix-redirect/compare/v1.4.54...v1.4.55) (2021-11-10)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.4.7 ([#358](https://github.com/adobe/helix-redirect/issues/358)) ([107a10d](https://github.com/adobe/helix-redirect/commit/107a10dc8322c5def815f29b11291eb62b187f49))

## [1.4.54](https://github.com/adobe/helix-redirect/compare/v1.4.53...v1.4.54) (2021-10-25)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.4.5 ([#351](https://github.com/adobe/helix-redirect/issues/351)) ([069ae43](https://github.com/adobe/helix-redirect/commit/069ae4394c4d2d8234cba01da8bfc6dc967540ee))

## [1.4.53](https://github.com/adobe/helix-redirect/compare/v1.4.52...v1.4.53) (2021-10-21)


### Bug Fixes


test/index.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
process.env.HELIX_FETCH_FORCE_HTTP1 = 'true';

const assert = require('assert');
const querystring = require('querystring');
const { main } = require('../src/index.js');
const { setupPolly } = require('./utils.js');

function retrofit(fn) {
  const resolver = {
    createURL({ package, name, version }) {
      return new URL(`https://adobeioruntime.net/api/v1/web/helix/${package}/${name}@${version}`);
    },
  };
  return async (params = {}, env = {}) => {
    const resp = await fn({
      url: `https://helix.com/redirect?${querystring.encode(params)}`,
      headers: new Map(Object.entries(params.__ow_headers || {})),
    }, {
      resolver,
      env,
    });
    return {
      statusCode: resp.status,
      body: String(resp.body),
      headers: [...resp.headers.keys()].reduce((result, key) => {
        // eslint-disable-next-line no-param-reassign
        result[key] = resp.headers.get(key);
        return result;
      }, {}),
    };
  };
}

const index = retrofit(main);

describe('Index Tests', () => {
  setupPolly({
    recordIfMissing: false,
  });

  it('400 when params missing', async () => {
    const result = await index({});
    assert.equal(result.statusCode, 400);
  });

  it('204 when no path provided', async function test() {
    const { server } = this.polly;

    server
      .get('https://raw.githubusercontent.com/adobe/theblog/branch1/:path')
      .intercept((req, res) => {
        res.status(200).send(`
redirects:
  - from: (.*).php
    to: $1.html
    type: temporary
      `);
      });

    const result = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'branch1',
    });
    assert.equal(result.statusCode, 204);
  });

  it('204 for non existing redirect.yaml', async function test() {
    const { server } = this.polly;

    server
      .get('https://raw.githubusercontent.com/adobe/theblog/branch2/helix-redirects.yaml')
      .intercept((req, res) => {
        res.status(404).send();
      });

    const result = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'branch2',
    });
    assert.equal(result.statusCode, 204);
  });

  it('204 when non-matching path provided', async function test() {
    const { server } = this.polly;

    server
      .get('https://raw.githubusercontent.com/adobe/theblog/branch2/:path')
      .intercept((req, res) => {
        res.status(200).send(`
redirects:
  - from: (.*).php
    to: $1.html
    type: temporary
      `);
      });

    const result = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'branch2',
      path: '/do-not-redirect-me',
    });
    assert.equal(result.statusCode, 204);
  });

  it('301 for PHP on private repo', async function test() {
    const { server } = this.polly;

    server
      .get('https://raw.githubusercontent.com/adobe/theblog/branch3p/helix-redirects.yaml')
      .intercept((req, res) => {
        if (req.headers.authorization !== 'token foobar') {
          res.status(404).send();
          return;
        }
        res.status(200).send(`
redirects:
  - from: (.*).php
    to: $1.html
      `);
      });

    const result = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'branch3p',
      path: '/test.php',
      __ow_headers: {
        'x-github-token': 'foobar',
      },
    });
    assert.equal(result.statusCode, 301);
  });

  it('301 for PHP', async function test() {
    const { server } = this.polly;

    server
      .get('https://raw.githubusercontent.com/adobe/theblog/branch3/:path')
      .intercept((req, res) => {
        res.status(200).send(`
redirects:
  - from: (.*).php
    to: $1.html
      `);
      });

    const result = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'branch3',
      path: '/test.php',
    });
    assert.equal(result.statusCode, 301);
  });

  it('301 for dynamic redirect with json', async function test() {
    const { server } = this.polly;
    const data = {
      total: 1,
      offset: 0,
      limit: 1,
      data: [
        {
          Source: '/tag/coronavirus/',
          Destination: 'https://blog.adobe.com/en/topics/covid-19.html',
          Notes: '',
        },
      ],
    };

    server
      .get('https://raw.githubusercontent.com/adobe/theblog/branch7/:path')
      .intercept((req, res) => {
        res.status(200).send(`
redirects:
  - https://master--theblog--adobe.hlx.page/redirects.json
      `);
      });

    server
      .get('https://master--theblog--adobe.hlx.page/redirects.json')
      .intercept((req, res) => {
        res.status(200).send(JSON.stringify(data));
      });

    const result = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'branch7',
      path: '/tag/coronavirus/',
    });
    assert.equal(result.statusCode, 301);
    assert.equal(result.headers.location, 'https://blog.adobe.com/en/topics/covid-19.html');
  });

  it('Temp Redirect', async function test() {
    const { server } = this.polly;

    server
      .get('https://raw.githubusercontent.com/adobe/theblog/branch4/:path')
      .intercept((req, res) => {
        res.status(200).send(`
redirects:
  - from: (.*).php
    to: $1.html
    type: temporary
      `);
      });

    const result = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'branch4',
      path: '/test.php',
    });
    assert.equal(result.statusCode, 302);
  });

  it('Internal Redirect', async function test() {
    const { server } = this.polly;

    server
      .get('https://raw.githubusercontent.com/adobe/theblog/branch5/:path')
      .intercept((req, res) => {
        res.status(200).send(`
redirects:
  - from: (.*).php
    to: $1.html
    type: internal
      `);
      });

    const result = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'branch5',
      path: '/test.php',
    });
    assert.equal(result.statusCode, 307);
  });

  it('Feed Redirect', async function test() {
    const { server } = this.polly;

    server
      .get('https://raw.githubusercontent.com/adobe/theblog/branch6/:path')
      .intercept((req, res) => {
        res.status(200).send(`
redirects:
  - from: .*/feed/$
    to: /feed.xml
    type: temporary
  - from: .*/feed$
    to: /feed.xml
    type: temporary
      `);
      });

    const result1 = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'branch6',
      path: '/tags/news/feed',
    });
    assert.equal(result1.statusCode, 302);

    const result2 = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'branch6',
      path: '/tags/news/feed/',
    });
    assert.equal(result2.statusCode, 302);
  });
});


test/post-deploy.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
/* eslint-disable no-unused-expressions */
const chai = require('chai');
const chaiHttp = require('chai-http');
const { createTargets } = require('./post-deploy-utils.js');

chai.use(chaiHttp);
const { expect } = chai;

createTargets().forEach((target) => {
  describe(`Post-Deploy Tests (${target.title()}) #online`, () => {
    before(function beforeAll() {
      if (!target.enabled()) {
        this.skip();
      }
    });

    it('No Redirect', async () => {
      const qs = '?owner=trieloff&repo=helix-demo&ref=528fd4692b6e4cd47ee9a11a133e7c6728b51fe5&path=test.md';
      const url = `${target.urlPath()}${qs}`;
      // eslint-disable-next-line no-console
      console.log(`Trying ${target.host()}${url}`);
      await chai
        .request(target.host())
        .get(url)
        .redirects(0)
        .then((response) => {
          expect(response).to.have.status(204);
        })
        .catch((e) => {
          throw e;
        });
    }).timeout(20000);

    it('302 Redirect', async () => {
      const qs = '?owner=trieloff&repo=helix-demo&ref=528fd4692b6e4cd47ee9a11a133e7c6728b51fe5&path=test.php';
      const url = `${target.urlPath()}${qs}`;
      // eslint-disable-next-line no-console
      console.log(`Trying ${target.host()}${url}`);
      await chai
        .request(target.host())
        .get(url)
        .redirects(0)
        .then((response) => {
          expect(response).to.have.status(302);
        })
        .catch((e) => {
          throw e;
        });
    }).timeout(20000);

    it('/feed Redirect', async () => {
      const qs = '?owner=adobe&repo=theblog&ref=34e880537c4bb787cdc6df0b71fcf76cc496bca5&path=/tags/news/feed';
      const url = `${target.urlPath()}${qs}`;
      // eslint-disable-next-line no-console
      console.log(`Trying ${target.host()}${url}`);
      await chai
        .request(target.host())
        .get(url)
        .redirects(0)
        .then((response) => {
          expect(response).to.have.status(302);
        })
        .catch((e) => {
          throw e;
        });
    }).timeout(50000);

    it('/feed/ Redirect', async () => {
      const qs = '?owner=adobe&repo=theblog&ref=34e880537c4bb787cdc6df0b71fcf76cc496bca5&path=/tags/news/feed/';
      const url = `${target.urlPath()}${qs}`;
      // eslint-disable-next-line no-console
      console.log(`Trying ${target.host()}${url}`);
      await chai
        .request(target.host())
        .get(url)
        .redirects(0)
        .then((response) => {
          expect(response).to.have.status(302);
        })
        .catch((e) => {
          throw e;
        });
    }).timeout(20000);
  });
});


test/post-deploy-utils.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-disable max-classes-per-file */
const packjson = require('../package.json');
require('dotenv').config();

class OpenwhiskTarget {
  constructor(opts = {}) {
    Object.assign(this, {
      namespace: 'helix',
      package: 'helix-services',
      name: packjson.name.replace('@adobe/helix-', ''),
      version: String(packjson.version),
    }, opts);
    if (process.env.CI && process.env.CIRCLE_BUILD_NUM && process.env.CIRCLE_BRANCH !== 'main' && !opts.version) {
      this.version = `ci${process.env.CIRCLE_BUILD_NUM}`;
    }
  }

  // eslint-disable-next-line class-methods-use-this
  title() {
    return 'OpenWhisk';
  }

  // eslint-disable-next-line class-methods-use-this
  host() {
    return 'https://adobeioruntime.net';
  }

  urlPath() {
    return `/api/v1/web/${this.namespace}/${this.package}/${this.name}@${this.version}`;
  }

  // eslint-disable-next-line class-methods-use-this
  enabled() {
    return true;
  }
}

class AWSTarget extends OpenwhiskTarget {
  // eslint-disable-next-line class-methods-use-this
  title() {
    return 'AWS';
  }

  // eslint-disable-next-line class-methods-use-this
  host() {
    return `https://${process.env.HLX_AWS_API}.execute-api.${process.env.HLX_AWS_REGION}.amazonaws.com`;
  }

  urlPath() {
    return `/${this.package}/${this.name}/${this.version}`;
  }

  // eslint-disable-next-line class-methods-use-this
  enabled() {
    return process.env.HLX_AWS_API && process.env.HLX_AWS_REGION;
  }
}

class GoogleTarget extends OpenwhiskTarget {
  title() {
    return 'Google';
  }

  host() {
    return `https://${process.env.HLX_GOOGLE_REGION}-${process.env.HLX_GOOGLE_PROJECT_ID}.cloudfunctions.net`;
  }

  urlPath() {
    return `/${this.package}--${this.name}_${this.version.replace(/\./g, '_')}`;
  }

  enabled() {
    return process.env.HLX_GOOGLE_PROJECT_ID;
  }
}

const ALL_TARGETS = [
  OpenwhiskTarget,
  AWSTarget,
  GoogleTarget,
];

function createTargets(opts) {
  return ALL_TARGETS.map((TargetClass) => new TargetClass(opts));
}

module.exports = {
  OpenwhiskTarget,
  AWSTarget,
  createTargets,
};


src/index.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const { Response } = require('@adobe/helix-fetch');
const wrap = require('@adobe/helix-shared-wrap');
const { logger } = require('@adobe/helix-universal-logger');
const { wrap: status } = require('@adobe/helix-status');
const { RedirectConfig } = require('@adobe/helix-shared-config');

/**
 * This is the main function
 * @param {Request} req The request
 */
async function main(req) {
  const { searchParams } = new URL(req.url);
  const owner = searchParams.get('owner');
  const repo = searchParams.get('repo');
  const ref = searchParams.get('ref');
  const path = searchParams.get('path');
  const githubToken = req.headers.get('x-github-token');
  if (!(owner && repo && ref)) {
    return new Response('missing owner, repo, or ref', {
      status: 400,
    });
  }

  const opts = {};
  if (githubToken) {
    opts.headers = {
      authorization: `token ${githubToken}`,
    };
  }

  const config = await new RedirectConfig()
    .withRepo(owner, repo, ref, opts)
    .init();

  const match = await config.match(path);

  if (match && match.type === 'temporary') {
    return new Response('moved temporarily', {
      status: 302,
      headers: {
        Location: match.url,
      },
    });
  } else if (match && match.type === 'permanent') {
    return new Response('moved permanently', {
      status: 301,
      headers: {
        'Cache-Control': 'max-age=30000000',
        Location: match.url,
      },
    });
  } else if (match && match.type === 'internal') {
    return new Response('moved internally', {
      status: 307,
      headers: {
        'HLX-Refetch': 'yes',
        Location: match.url,
      },
    });
  }
  return new Response('No redirect', {
    status: 204, // no content
  });
}

module.exports.main = wrap(main)
  .with(status)
  .with(logger.trace)
  .with(logger);
